src/util/serialize.cpp

```diff
@@ -120,121 +120,152 @@ std::string deSerializeString32(std::istream &is)
 }
 
 ////
-//// JSON
+//// JSON-like strings
 ////
 
 std::string serializeJsonString(const std::string &plain)
 {
-	std::ostringstream os(std::ios::binary);
-	os << "\"";
+	std::string tmp;
+
+	tmp.reserve(plain.size() + 2);
+	tmp.push_back('"');
 
 	for (char c : plain) {
 		switch (c) {
 			case '"':
-				os << "\\\"";
+				tmp.append("\\\"");
 				break;
 			case '\\':
-				os << "\\\\";
-				break;
-			case '/':
-				os << "\\/";
+				tmp.append("\\\\");
 				break;
 			case '\b':
-				os << "\\b";
+				tmp.append("\\b");
 				break;
 			case '\f':
-				os << "\\f";
+				tmp.append("\\f");
 				break;
 			case '\n':
-				os << "\\n";
+				tmp.append("\\n");
 				break;
 			case '\r':
-				os << "\\r";
+				tmp.append("\\r");
 				break;
 			case '\t':
-				os << "\\t";
+				tmp.append("\\t");
 				break;
 			default: {
 				if (c >= 32 && c <= 126) {
-					os << c;
+					tmp.push_back(c);
 				} else {
-					u32 cnum = (u8)c;
-					os << "\\u" << std::hex << std::setw(4)
-						<< std::setfill('0') << cnum;
+					// We pretend that Unicode codepoints map to bytes (they don't)
+					u8 cnum = static_cast<u8>(c);
+					tmp.append("\\u00");
+					tmp.push_back(hex_chars[cnum >> 4]);
+					tmp.push_back(hex_chars[cnum & 0xf]);
 				}
 				break;
 			}
 		}
 	}
 
-	os << "\"";
-	return os.str();
+	tmp.push_back('"');
+	return tmp;
+}
+
+static void deSerializeJsonString(std::string &s)
+{
+	assert(s.size() >= 2);
+	assert(s.front() == '"' && s.back() == '"');
+
+	size_t w = 0; // write index
+	size_t i = 1; // read index
+	const size_t len = s.size() - 1; // string length with trailing quote removed
+
+	while (i < len) {
+		char c = s[i++];
+		assert(c != '"');
+
+		if (c != '\\') {
+			s[w++] = c;
+			continue;
+		}
+
+		if (i >= len)
+			throw SerializationError("JSON string ended prematurely");
+		char c2 = s[i++];
+		switch (c2) {
+			case 'b':
+				s[w++] = '\b';
+				break;
+			case 'f':
+				s[w++] = '\f';
+				break;
+			case 'n':
+				s[w++] = '\n';
+				break;
+			case 'r':
+				s[w++] = '\r';
+				break;
+			case 't':
+				s[w++] = '\t';
+				break;
+			case 'u': {
+				int hexnumber;
+				char hexdigits[4 + 1];
+
+				if (i + 3 >= len)
+					throw SerializationError("JSON string ended prematurely");
+				memcpy(hexdigits, &s[i], 4);
+				i += 4;
+				hexdigits[4] = 0;
+
+				std::istringstream tmp_is(hexdigits, std::ios::binary);
+				tmp_is >> std::hex >> hexnumber;
+				// Note that this does not work for anything other than ASCII
+				// but these functions do not actually interact with real JSON input.
+				s[w++] = (char) hexnumber;
+				break;
```

Comment:
util/hex.h is missing a good interface for doing this. But for now could you do: (haven't tested this)

```suggestion
				u16 hexnumber = 0;
				for (int k = 0; k < 4; k++) {
					if (i >= len)
						throw SerializationError("JSON string ended prematurely");
					unsigned char tmpchar;
					if (!hex_digit_decode(s[i++], tmpchar))
						throw SerializationError("JSON string invalid unicode");
					hexnumber <<= 4;
					hexnumber |= tmpchar;
				}
				s[w++] = (char) hexnumber;
``````

Reply:
I think I prefer looks over performance here :wink:

Reply:
Actually, this both looks good and will be faster:
```cpp
// length check
unsigned char c[4] = {};
for (int j = 0; i < 4; i++))
    hex_digit_decode(s[i+j], &c[j]);
u32 hexnumber = c[3] << 12 | c[2] << 8 | c[1] << 4 | c[0];
// the rest
```

Reply:
You and me have a very different aesthetic sense! I avoid `stringstream` like the plague. Ancient C++ IO APIs are wildly inefficient.

But yes, that is much better. Note that the two i's on the for line should be j.

Reply:
The order of the hex digits also needs to be reversed. c[0] will have the most-significant digit.

Reply:
Yes. I stared at it for about 15 minutes because I made three mistakes quickly writing this down...